daffodil-io/src/main/scala/org/apache/daffodil/io/InputSource.scala

```diff
@@ -22,6 +22,9 @@ import java.nio.ByteBuffer
 import scala.collection.mutable.ArrayBuffer
 
 import org.apache.daffodil.exceptions.Assert
+import org.apache.daffodil.exceptions.ThinThrowable
+
+case class BacktrackingException() extends Exception with ThinThrowable
```

Comment:
Definitely agreed. I think it's also important to point out that the act of backtracking (i.e releasePosition) itself does not cause an error. The error only occurs when someone tries to read from that point. I could imagine a case where you try to parse this huge optional thing, we eventually find out it's optional and throw away the hole thing, and then the parse is done. We can actually return a valid infoset in that case, even though we've backtracked too far to read anymore data.

Note that if someone tried to use this for a second parse (i.e. stream) then would hit the backtrack error. This suggests that the comment I made above about marking an input source as invalid vs throw a backtrack error the bucket is negative, the right solution is throw when negative since the input source wouldn't be invalid in this case.